### PR TITLE
Boost dependences is removed

### PR DESCRIPTION
**Pull request by dvenum**
1. I added -std=c++11 flag for std::rand support. Tell me, if it matter.

2. EuclideanIndexTest::test_get_nns_by_vector failed, see below. It's in reverse order, but AngularIndexTest::, passed ok. I don't know, how I can receive [1,0], if it include sort call before return.
annoylib.h:
537     sort(nns_dist.begin(), nns_dist.end());
538     for (size_t i = 0; i < nns_dist.size() && result->size() < n; i++) {
539       result->push_back(nns_dist[i].second);
540     }

FAIL: test_get_nns_by_vector (annoy_test.EuclideanIndexTest)

Traceback (most recent call last):
  File "/home/dvenum/devl/annoy/test/annoy_test.py", line 112, in test_get_nns_by_vector
    self.assertEqual(i.get_nns_by_vector([3,3], 2), [1, 0])
AssertionError: Lists differ: [0, 1] != [1, 0]
First differing element 0:
0
1
<div>
-  [0, 1] <br/>
+  [1, 0] <br/>
<br/>
3. If I right, valgrin have clear output, only usually python message. It seems to true, because a small new memory management was delegated to python.
</div>

### CODE REVIEW

**Comment by erikbern** [comment]
Looks good.

The sort on line 537 sorts by the first element of each pair (the distance) which is smaller for element 1 than for element 0.

Element 1 should have squared distance 2 and element 0 should have squared distance 3.

**Comment by dvenum** [comment]
It seems to my bad. nns_dist contain:
8 0
13 1

But for angular test: 
2 0
2 1
2 2

**Comment by erikbern** [comment]
Look at https://github.com/spotify/annoy/blob/master/test/annoy_test.py#L105

`nn_dist` should contain
1 1
2 0

(sorry I incorrectly said 2 and 3 in my previous comment, that was not right)

File: src/annoymodule.cc [hunk 1]
```diff
@@ -14,87 +14,490 @@
...
+    ai = static_cast<AI*>(self->ptr);
+    res = ai->load_py(filename);
+    break;
+  case 'e':
+    ei = static_cast<EI*>(self->ptr);
+    res = ei->load_py(filename);
+    break;
+  }
+  if (!res) {
+    PyErr_SetString(PyExc_IOError,"load error");
```
Comment (a1k0n) [inline src/annoymodule.cc]: It should be safe to use PyErr_SetFromErrno(PyExc_IOError); here, as errno should be preserved from the load_py call. I think.

File: src/annoymodule.cc [hunk 2]
```diff
@@ -14,87 +14,490 @@
...
+  0,						 /* tp_getset */
+  0,						 /* tp_base */
+  0,						 /* tp_dict */
+  0,						 /* tp_descr_get */
+  0,						 /* tp_descr_set */
+  0,						 /* tp_dictoffset */
+  (initproc)py_an_init,		 /* tp_init */
+  0,						 /* tp_alloc */
+  py_an_new,				 /* tp_new */
+};
```
Comment (a1k0n) [inline src/annoymodule.cc]: would you mind expanding the tabs to spaces above here?

File: src/annoymodule.cc [hunk 3]
```diff
@@ -14,87 +14,490 @@
...
+    ai = static_cast<AI*>(self->ptr);
+    res = ai->save_py(filename);
+    break;
+  case 'e':
+    ei = static_cast<EI*>(self->ptr);
+    res = ei->save_py(filename);
+    break;
+  }
+  if (!res) {
+    PyErr_SetString(PyExc_IOError,"load error");
```
Comment (a1k0n) [inline src/annoymodule.cc]: `PyErr_SetFromErrno(PyExc_IOError);` here too, would be more informative.

**Comment by a1k0n** [comment]
If AI and EI were subclasses with virtual member functions, you could save a whole lot of dispatch code here.

**Comment by erikbern** [comment]
As @a1k0n pointed out making AI and EI subclasses with virtual functions seems like the easiest way to avoid the switching and static casting.

Another way is to move the whole switching to compile time and compile two different python extensions. But that's a bit more complex and the performance gains are pretty irrelevant

**Comment by dvenum** [comment]
As @erikbern pointed me, it will similar to:
// annoy python object
template<typename S, typename T, typename Distance>
typedef struct {
    PyObject_HEAD
    int f;
    char ch_metric;
    AnnoyIndexPython<S, T, Distance >* ptr;
} py_annoy;

I was a little confusing with templates exposing. It was first working code with rewriting purpose.

File: src/annoymodule.cc (hunk 2 above)
Comment (dvenum) [inline src/annoymodule.cc]: It's done, sorry.

**Comment by dvenum** [comment]
Critical bugfix in c02bb89. Input vector was always zero.
test_get_nns_by_vector passed.

**Comment by erikbern** [comment]
I don't understand why it solves the problem but great that it works now!

**Comment by dvenum** [comment]
push_back was used me for vector object when it convering from input python list. New items was added to end, but it using by index.
```c
    vector<T> w(this->_f); <<-- start length no zero
    for (int z = 0; z < PyList_Size(v) && z < this->_f; z++) {   <<-- check border also
        PyObject *pf = PyList_GetItem(v,z);
        w[z] = PyFloat_AsDouble(pf);             <<-- correct way
    }
```

**Comment by erikbern** [comment]
oh ok, makes sense! of course. thanks for the explanation!

**Comment by dvenum** [comment]
I have same problem with templates.

As below, object py_annoy should be allocated when I don't know, what is object required. No metrics received yet. ptr should be a pointer to an universal base class.
```c
// annoy python object
template<typename S, typename T, typename Distance>
struct py_annoy_ { 
  PyObject_HEAD
  int f;
  char ch_metric;
  void* ptr;
  AnnoyIndexPython<S, T, Distance >* ptr;
};

static PyObject *
py_an_new(PyTypeObject *type, PyObject *args, PyObject *kwds) {
  py_annoy *self;

  self = (py_annoy *)type->tp_alloc(type, 0);
  if (self != NULL) {
    self->f = 0;
    self->ch_metric = 0;
    self->ptr = NULL;
  }

  return (PyObject *)self;
}
```

Boost determine class template, which may wrap C++ class and expose him. It's a bit complex way for this task.
I think about some tricks in C-style. It'll ok when we will have few wrapper functions if no code will duplicated.

**Comment by erikbern** [comment]
Ok, that's fine, I can take a look at it later.

Can you remove the "-std=c++11" flag though? Travis-CI doesn't like it for some reason

**Comment by dvenum** [comment]
<random> lib exists only for c++11. I can include another library as part of sources if nothing to more will be.

May you try something, similar to it in .travis.yml?
```bash
install:
  - sudo apt-get install -qq gcc-4.8 g++-4.8 
# We want to compile with g++ 4.8 when rather than the default g++
  - sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-4.8 90
```

**Comment by erikbern** [comment]
Which lib exists only for c++11? I think `#include <random>` should work for older versions too

**Comment by dvenum** [comment]
As you can see on the head, only C++11 marker exists: 
http://www.cplusplus.com/reference/random/

Old C++ have only rand() function from stdlib.

**Comment by erikbern** [comment]
Ok maybe we just have to include something like this: http://phoxis.org/2013/05/04/generating-random-numbers-from-normal-distribution-in-c/

**Comment by dvenum** [comment]
If using rand() call is correct, I can rewrite it, no problem. Or I can include a bit more complex library.
Travis-CI uses gcc-4.6, that very old.

**Comment by erikbern** [comment]
I can fix. Let's merge this now!

**Comment by erikbern** [comment]
It was pretty easy: https://github.com/spotify/annoy/pull/52

**Pull request merged**